lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -173,6 +173,73 @@ struct StorageGetOpLowering : public OpConversionPattern<arc::StorageGetOp> {
   }
 };
 
+struct AssertOpLowering : public OpConversionPattern<arc::AssertOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AssertOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+
+    rewriter.replaceOpWithNewOp<scf::IfOp>(
+        op, adaptor.getProperty(),
+        [&](auto &builder, auto loc) {
+          // if property is true then assert will not execute
+          builder.template create<scf::YieldOp>(loc);
+        },
+        [&](auto &builder, auto loc) {
+          // If the condition doesn't hold then print the message and abort
+          auto message = op.getMsg().has_value() ? op.getMsg().value() : "";
+          auto module = op->getParentOfType<ModuleOp>();
+
+          LLVM::LLVMFuncOp putsFunc;
+          LLVM::GlobalOp globalOp;
+
+          if (!message.empty()) {
+            // Only use puts if there's a message provided
+            OpBuilder::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPointToStart(module.getBody());
+            putsFunc = lookupOrCreateFn(
+                module, "puts", LLVM::LLVMPointerType::get(getContext()),
+                LLVM::LLVMVoidType::get(getContext()), false);
+            // Convert the string message to a Twine.
+            llvm::StringRef msgRef(message);
+
+            // Get the integer type with the correct bit width.
+            Type int8Type = IntegerType::get(op.getContext(), 8);
+
+            // Create a global constant with the message.
+            globalOp = rewriter.create<LLVM::GlobalOp>(
+                loc, LLVM::LLVMArrayType::get(int8Type, message.size()),
+                /*isConstant=*/true, LLVM::Linkage::Internal, "message",
+                rewriter.getStringAttr(msgRef));
+          }
+
+          if (!message.empty()) {
+            Value globalString =
+                rewriter.create<LLVM::AddressOfOp>(loc, globalOp);
+
+            builder.template create<LLVM::CallOp>(loc, putsFunc,
+                                                  ValueRange{globalString});
+          }
+
+          // Abort after printing the message
+          auto abortFunc = module.lookupSymbol<LLVM::LLVMFuncOp>("abort");
+          if (!abortFunc) {
+            // if there's no abort create it
+            OpBuilder::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPointToStart(module.getBody());
+            auto abortFuncTy = LLVM::LLVMFunctionType::get(
+                LLVM::LLVMVoidType::get(getContext()), {});
+            abortFunc =
+                rewriter.create<LLVM::LLVMFuncOp>(loc, "abort", abortFuncTy);
+          }
+          builder.template create<LLVM::CallOp>(loc, abortFunc, std::nullopt);
```

Comment:
There exists a utility header (`FunctionCallUtils`) in the LLVM dialect with a function `lookupOrCreateFn`. I think it does exactly that, or am I missing something?

Reply:
Nope, you're right. I did it just for experimental reasons and I forgot to change it. Sorry.